_drafts/unittests-cpputest.md

```diff
@@ -0,0 +1,1251 @@
+
+{:.no_toc}
+
+### Stubs
+
+These are used when the implementation of specific functions or their return
+values do not matter to the module under test. They are primarily used to fix
+the linker's `ld: symbol(s) not found` errors. These should generally have only
+a return statement that always returns `true`, `false`, `0`, `NULL`, or whatever
+makes sense in the context of the module.
+
+If there is anything more complex than a return statement, consider implemented
+a **Fake** instead.
+
+Examples:
+
+- Hardware or peripheral initialization functions since they have little
+  relevance in testing on the host (x86 machine).
+- A time module which returns the time of day (just return a random time).
+- Mutex stubs when the locking or unlocking isn't being tested. (shown below)
+
+```c
+#include "mutex/mutex.h"
+
+Mutex *mutex_create(void) {
+  return NULL;
+}
+
+void mutex_lock(Mutex *mutex) {
+  return;
+}
+
+void mutex_unlock(Mutex *mutex) {
+  return;
+}
+```
+
+{:.no_toc}
+
+### Fakes
+
+A **Fake** is commonly used in firmware where it is impractical to use the real
+implementation for reasons such as:
+
+- It requires specific hardware (flash chip, peripherals, LED's, etc.)
+
+Examples:
+
+- A time module which accepts an initial UTC time at the start of the test and
+  allows incrementing the time manually.
+- A mutex module which checks at the end of the test that all mutexes were
+  properly unlocked (example provided later in the post).
+- A RAM based NOR flash implementation, which flips bits from `1` to `0` when
+  written to and requires "erasing" (flipping bits back to `1`) before writing
+  new data.
+- A RAM based key-value store (example shown below).
+
+```c
+#include "kv_store.h"
+
+#include <inttyptes.h>
+
+typedef struct {
+  char *key;
+  void *val;
+  uint32_t len;
+} KvEntry;
+
+static KvEntry s_kv_store[256];
+
+bool kv_store_write(const char *key, const void *val, uint32_t len) {
+  // Write value into RAM key-value store
+}
+
+uint32_t kv_store_read(const char *key, void *buf, uint32_t buf_len) {
+  // Read value from RAM key-value store
+}
+
+bool kv_store_delete(const char *key) {
+  // Delete value from RAM key-value store
+}
+```
+
+{:.no_toc}
+
+### Mocks
+
+Mocks are incredibly useful if you want to declare each and every return value
+of a given function. Using many mocks in a single unit test is also the easiest
+way to test every single code path of the module under test, as you can force
+any function to return error codes, NULL values, and invalid pointers.
+
+These are the most powerful, provide the programmer the most control, and
+isolate the module under test the best, but they are also the most cumbersome
+and verbose to use, as every return value has to be pre-programmed.
+
+Common mocking frameworks include:
+
+- [CppUMock](https://cpputest.github.io/mocking_manual.html)
+- [CMock](http://www.throwtheswitch.org/cmock)
+- [fff](https://github.com/meekrosoft/fff) NOTE: Although the name suggests it
```

Comment:
The formatting is a bit awkward, maybe italicize the note and add an emdash? "-"

Reply:
I really struggled with the formatting. Italicized is a good idea.